File: elixir/word-count/word_count.exs
defmodule Words do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t) :: map
  def count(sentence) do
    String.downcase(sentence)
    |> String.replace(~r/[!|&|@|\$|%|\^|&|,]+/, "")
    |> String.split([" ", "_"])
    |> Enum.filter(fn(word) -> String.match?(word, ~r/\w+/) end)
    |> Enum.reduce(Map.new, fn(word, map) -> Map.update(map, word, 1, &(&1 + 1)) end)
  end

end

File: typescript/word-count/word-count.ts
class Words {
  count(input: string): Map<string, number> {
    const words = input.toLowerCase().replace(/\n/m, " ").replace(/\t/m, " ").split(" ")
    const result = new Map<string, number>()

    for (const word of words) {
      if (word === "") {
        continue
      }

      const count = result.get(word) || 0
      result.set(word.trim(), count + 1)
    }

    return result
  }
}

export default Words
